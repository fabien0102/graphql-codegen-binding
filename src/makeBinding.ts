import {
  DocumentNode,
  Kind,
  ObjectTypeDefinitionNode,
  OperationTypeDefinitionNode,
  parse,
} from 'graphql'
import { GraphQLSchema } from 'graphql/type/schema'
import { buildASTSchema } from 'graphql/utilities/buildASTSchema'
import { generators } from './generators'
import { GeneratorType } from './types'

/**
 * The schema contains incompatible characters sometimes, e.g.
 * data types in comments are emphasized with "`", which represents
 * template strings in ES2015 and TypeScript. This function
 * replaces those characters with sane defaults.
 *
 * @param schema {String} The serialized schema
 * @returns {String}
 *
 */
const sanitizeSchema = (schema: string) => schema.replace(/\`/g, "'")

export function makeBinding(
  schema: string,
  generatorName: GeneratorType,
): string {
  const generator =
    generators[generatorName] || require(generatorName).generator
  if (!generator) {
    throw new Error(`Generator '${generator}' could not be found. Available generators:
${Object.keys(generators)
      .map(k => `'${k}`)
      .join(', ')}`)
  }

  schema = sanitizeSchema(schema)

  const document: DocumentNode = parse(schema, { noLocation: true })
  const ast: GraphQLSchema = buildASTSchema(document)

  // Create types
  const typeNames = Object.keys(ast.getTypeMap())
    .filter(typeName => !typeName.startsWith('__'))
    .filter(typeName => typeName !== (ast.getQueryType() as any).name)
    .filter(
      typeName =>
        ast.getMutationType()
          ? typeName !== (ast.getMutationType()! as any).name
          : true,
    )
    .filter(
      typeName =>
        ast.getSubscriptionType()
          ? typeName !== (ast.getSubscriptionType()! as any).name
          : true,
    )
    .sort(
      (a, b) =>
        (ast.getType(a) as any).constructor.name <
        (ast.getType(b) as any).constructor.name
          ? -1
          : 1,
    )

  // Special case 4: header
  const generatedClass = [generator.Header(schema)]

  // Process all types
  generatedClass.push(
    ...typeNames.map(typeName => {
      const type = ast.getTypeMap()[typeName]
      return generator[type.constructor.name]
        ? generator[type.constructor.name](type)
        : null
    }),
  )

  // Special case 1: generate schema interface
  if (generator.SchemaType) {
    generatedClass.push(
      generator.SchemaType(
        ast.getQueryType(),
        ast.getMutationType(),
        ast.getSubscriptionType(),
      ),
    )
  }

  // Special case 2: generate root field interfaces
  if (generator.RootType) {
    generatedClass.push(generator.RootType(ast.getQueryType()))
    if (ast.getMutationType()) {
      generatedClass.push(generator.RootType(ast.getMutationType()!))
    }
  }

  // Special case 5: subscription type
  if (generator.SubscriptionType) {
    if (ast.getSubscriptionType()) {
      generatedClass.push(
        generator.SubscriptionType(ast.getSubscriptionType()!),
      )
    }
  }

  // Special case 3: the main method
  generatedClass.push(
    generator.Main(
      ast.getQueryType(),
      ast.getMutationType(),
      ast.getSubscriptionType(),
    ),
  )

  return generatedClass.filter(r => r).join('\n\n')
}
