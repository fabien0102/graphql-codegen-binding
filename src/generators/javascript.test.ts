import { makeBinding } from '../makeBinding'
import * as fs from 'fs'

const schema = fs.readFileSync(__dirname + '/fixtures/schema.graphql', 'utf-8')
test('typescript generator', () => {
  expect(makeBinding(schema, 'javascript')).toMatchSnapshot()
})
