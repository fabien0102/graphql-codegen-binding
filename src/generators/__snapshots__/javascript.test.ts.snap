// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript generator 1`] = `
"const { Binding: BaseBinding } = require('graphql-binding')
const { GraphQLResolveInfo } = require('graphql')

module.exports.Binding = class Binding extends BaseBinding {

  constructor({ schema, fragmentReplacements }) {
    super({ schema, fragmentReplacements });

    var self = this
    this.query = {
      posts(args, context, info) { 
        return self.delegate('query', 'posts', args, context, info)
      },
      users(args, context, info) { 
        return self.delegate('query', 'users', args, context, info)
      },
      post(args, context, info) { 
        return self.delegate('query', 'post', args, context, info)
      },
      user(args, context, info) { 
        return self.delegate('query', 'user', args, context, info)
      },
      postsConnection(args, context, info) { 
        return self.delegate('query', 'postsConnection', args, context, info)
      },
      usersConnection(args, context, info) { 
        return self.delegate('query', 'usersConnection', args, context, info)
      },
      node(args, context, info) { 
        return self.delegate('query', 'node', args, context, info)
      }
    }
      
    this.mutation = {
      createPost(args, context, info) { 
        return self.delegate('mutation', 'createPost', args, context, info)
      },
      createUser(args, context, info) { 
        return self.delegate('mutation', 'createUser', args, context, info)
      },
      updatePost(args, context, info) { 
        return self.delegate('mutation', 'updatePost', args, context, info)
      },
      updateUser(args, context, info) { 
        return self.delegate('mutation', 'updateUser', args, context, info)
      },
      deletePost(args, context, info) { 
        return self.delegate('mutation', 'deletePost', args, context, info)
      },
      deleteUser(args, context, info) { 
        return self.delegate('mutation', 'deleteUser', args, context, info)
      },
      upsertPost(args, context, info) { 
        return self.delegate('mutation', 'upsertPost', args, context, info)
      },
      upsertUser(args, context, info) { 
        return self.delegate('mutation', 'upsertUser', args, context, info)
      },
      updateManyPosts(args, context, info) { 
        return self.delegate('mutation', 'updateManyPosts', args, context, info)
      },
      updateManyUsers(args, context, info) { 
        return self.delegate('mutation', 'updateManyUsers', args, context, info)
      },
      deleteManyPosts(args, context, info) { 
        return self.delegate('mutation', 'deleteManyPosts', args, context, info)
      },
      deleteManyUsers(args, context, info) { 
        return self.delegate('mutation', 'deleteManyUsers', args, context, info)
      }
    }
      
    this.subscription = {
      post(args, context, infoOrQuery) { 
        return self.delegateSubscription('post', args, context, infoOrQuery)
      },
      user(args, context, infoOrQuery) { 
        return self.delegateSubscription('user', args, context, infoOrQuery)
      }
    }
  }

  delegate(operation, field, args, context, info) {
    return super.delegate(operation, field, args, context, info)
  }

  delegateSubscription(field, args, context, infoOrQuery) {
    return super.delegateSubscription(field, args, context, infoOrQuery)
  }
}"
`;
